### PR TITLE
Add support `bitemporal_callbacks`

### PR DESCRIPTION
**Pull request by Dooor**
## Summary

When using `ActiveRecord::Callbacks`, create callback is also called on `#update` and `#destroy` because there is an implicit record creation process in `ActievRecord::Bitemporal`.

This PR adds bitemporal callbacks like `after_bitemporal_create`, `after_bitemporal_update` and `after_bitemporal_destroy` that does not call create callback on `#update` and `#destroy`.

Below is a sample code.

### when use `ActiveRecord::Callbacks`

```rb
class Company < ActiveRecord::Base
  include ActiveRecord::Bitemporal

  [:after_create, :after_update, :after_destroy].each do |callback_method|
    send(callback_method, Proc.new { pp callback_method })
  end
end

company = Company.create!(name: "A")
# => :after_create

company.update!(name: "B")
# => :after_create
#    :after_create
#    :after_update

company.destroy!
# => :after_create
#    :after_destroy
```

### when use `ActiveRecord::Bitemporal::Callbacks`

```rb
class Company < ActiveRecord::Base
  include ActiveRecord::Bitemporal

  [:after_bitemporal_create, :after_bitemporal_update, :after_bitemporal_destroy].each do |callback_method|
    send(callback_method, Proc.new { pp callback_method })
  end
end

company = Company.create!(name: "A")
# => :after_bitemporal_create

company.update!(name: "B")
# => :after_bitemporal_update

company.destroy!
# => :after_bitemporal_destroy
```

### CODE REVIEW

**Comment by wata727** [comment]
It seems that CI has failed other than Rails 7.0.

File: activerecord-bitemporal.gemspec [hunk 1]
```diff
@@ -14,6 +14,7 @@ Gem::Specification.new do |spec|
   spec.description   = %q{Enable ActiveRecord models to be handled as BiTemporal Data Model.}
   spec.homepage      = "https://github.com/kufu/activerecord-bitemporal"
   spec.license       = "Apache 2.0"
+  spec.required_ruby_version = ">= 2.7.0"
```
Comment (wata727) [inline activerecord-bitemporal.gemspec:17]: This is should be worked as another PR.

File: spec/activerecord-bitemporal/callbacks_spec.rb [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    [:before_bitemporal_create, :after_bitemporal_create,
+     :before_bitemporal_update, :after_bitemporal_update,
+     :before_bitemporal_destroy, :after_bitemporal_destroy
+    ].each do |callback_method|
+      send(callback_method, Proc.new { |model| model.log << callback_method })
+    end
+  end
+end
+
+class EmployeeWithBitemporalCallbaksLog < Employee
```
Comment (wata727) [inline spec/activerecord-bitemporal/callbacks_spec.rb:21]:
```suggestion
class EmployeeWithBitemporalCallbacksLog < Employee
```

File: spec/activerecord-bitemporal/callbacks_spec.rb [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+      context 'without option `bitemporal_callbacks`' do
+        it { is_expected.to change(employee, :log).from([]).to(%i[before_bitemporal_create after_bitemporal_create]) }
+        it { is_expected.to change { employee.address.log }.from([]).to(%i[before_bitemporal_create after_bitemporal_create]) }
+      end
+
+      context 'employee with option `bitemporal_callbacks: false`' do
+        before { employee.bitemporal_option_merge!(bitemporal_callbacks: false) }
+        it { is_expected.not_to change(employee, :log) }
+        it { is_expected.to change { employee.address.log }.from([]).to(%i[before_bitemporal_create after_bitemporal_create]) }
```
Comment (wata727) [inline spec/activerecord-bitemporal/callbacks_spec.rb:69]: We should consider whether it makes sense for the parent's bitemporal option not to propagate to the child. For example, the `valid_datetime` is propagated. This is a lack of consistency.

```ruby
employee.bitemporal_option_merge!(valid_datetime: '2023/02/01')
employee.save!

employee.valid_from         # => 2023/02/01
employee.address.valid_from # => 2023/02/01
```

**Review by wata727 — COMMENTED** [review]
Nice work! Left some comments.

**Comment by osyo-manga** [comment]
@Dooor Great!
Can you adding example code in [CHANGELOG.md](https://github.com/kufu/activerecord-bitemporal/blob/master/CHANGELOG.md)?

File: spec/activerecord-bitemporal/callbacks_spec.rb (hunk 2 above)
Comment (Dooor) [reply spec/activerecord-bitemporal/callbacks_spec.rb]: 👍

File: spec/activerecord-bitemporal/callbacks_spec.rb (hunk 3 above)
Comment (Dooor) [reply spec/activerecord-bitemporal/callbacks_spec.rb:69]: Thanks for the comment. 
Other bitemporal options (e.g. force_update) are not propagated to children now. Actually, valid_datetime option is not propagated too when children were loaded before set valid_datetime option.

```rb
# if already loaded employee.address
employee.bitemporal_option_merge!(valid_datetime: '2023/02/01')
employee.save!

employee.valid_from         # => 2023/02/01 00:00:00
employee.address.valid_from # => 2023/02/01 15:34:10 (Time.current)
```

I think it's better to open a new issue if we need to fix this problem because it has a large impact.

File: activerecord-bitemporal.gemspec (hunk 1 above)
Comment (Dooor) [reply activerecord-bitemporal.gemspec]: I opened another PR for adding this change.
https://github.com/kufu/activerecord-bitemporal/pull/125

**Comment by Dooor** [comment]
@osyo-manga Thanks. I added example code in CHANGELOG.md.

File: lib/activerecord-bitemporal.rb [hunk 4]
```diff
@@ -123,6 +124,7 @@ def bitemporalize(
     extend ClassMethods
     include InstanceMethods
     include ActiveRecord::Bitemporal::Scope
+    include ActiveRecord::Bitemporal::Callbacks
```
Comment (wata727) [inline lib/activerecord-bitemporal.rb:127]: If it follows the same structure as upstream, it may be better to be included in `InstanceMethods`.

In the upstream, `Persistence` and `Callbacks` are included in the `Base`.
https://github.com/rails/rails/blob/v7.0.4.2/activerecord/lib/active_record/base.rb#L300
https://github.com/rails/rails/blob/v7.0.4.2/activerecord/lib/active_record/base.rb#L315

File: lib/activerecord-bitemporal/callbacks.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        bitemporal_option[:ignore_bitemporal_callbacks] != true
+      end
+    end
+
+    included do
+      define_model_callbacks :bitemporal_create
+      define_model_callbacks :bitemporal_update
+      define_model_callbacks :bitemporal_destroy
+
+      prepend Wrapper
```
Comment (wata727) [inline lib/activerecord-bitemporal/callbacks.rb:38]: Why do we need prepend here?

In the upstream, it seems that this is just be overridden.
https://github.com/rails/rails/blob/v7.0.4.2/activerecord/lib/active_record/callbacks.rb#L460-L462

Same for `ActiveRecord::Bitemporal::Persistence`.
https://github.com/kufu/activerecord-bitemporal/blob/05fcb5a3099e19e17fca1bc98f484c186be366a9/lib/activerecord-bitemporal/bitemporal.rb#L292-L298

File: lib/activerecord-bitemporal/patches.rb [hunk 6]
```diff
@@ -18,13 +18,14 @@ def klass
...
         refine Persistence do
+          COPY_BITEMPORAL_OPTION_EXCEPT_KEYS = %i[ignore_bitemporal_callbacks]
           def copy_bitemporal_option(src, dst)
             return unless [src.class, dst.class].all? { |klass|
               # NOTE: Can't call refine method.
               # klass.bi_temporal_model?
               klass.include?(ActiveRecord::Bitemporal)
             }
-            dst.bitemporal_option_merge! src.bitemporal_option
+            dst.bitemporal_option_merge! src.bitemporal_option.except(*COPY_BITEMPORAL_OPTION_EXCEPT_KEYS)
```
Comment (wata727) [inline lib/activerecord-bitemporal/patches.rb:28]: Why is only this option ignored?

File: spec/activerecord-bitemporal/callbacks_spec.rb [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_create after_bitemporal_create]) }
+      end
+
+      context 'employee with option `ignore_bitemporal_callbacks: true`' do
+        before { employee.bitemporal_option_merge!(ignore_bitemporal_callbacks: true) }
+        it { expect { subject }.not_to change(employee, :log) }
+        it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_create after_bitemporal_create]) }
+      end
+
+      context 'address with option `ignore_bitemporal_callbacks: true`' do
```
Comment (wata727) [inline spec/activerecord-bitemporal/callbacks_spec.rb:88]:
```suggestion
      context 'job title with option `ignore_bitemporal_callbacks: true`' do
```

File: spec/activerecord-bitemporal/callbacks_spec.rb [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+          it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
+        end
+
+        context 'employee with option `ignore_bitemporal_callbacks: true`' do
+          before { employee.bitemporal_option_merge!(ignore_bitemporal_callbacks: true) }
+          it { expect { subject }.not_to change(employee, :log) }
+          it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
+        end
+
+        context 'address with option `ignore_bitemporal_callbacks: true`' do
```
Comment (wata727) [inline spec/activerecord-bitemporal/callbacks_spec.rb:139]:
```suggestion
        context 'job title with option `ignore_bitemporal_callbacks: true`' do
```

File: spec/activerecord-bitemporal/callbacks_spec.rb [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
```
Comment (wata727) [inline spec/activerecord-bitemporal/callbacks_spec.rb:164]:
```suggestion
        context 'job title with option `ignore_bitemporal_callbacks: true`' do
```

File: spec/activerecord-bitemporal/callbacks_spec.rb [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_destroy after_bitemporal_destroy]) }
+      end
+
+      context 'employee with option `ignore_bitemporal_callbacks: true`' do
+        before { employee.bitemporal_option_merge!(ignore_bitemporal_callbacks: true) }
+        it { expect { subject }.not_to change(employee, :log) }
+        it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_destroy after_bitemporal_destroy]) }
+      end
+
+      context 'address with option `ignore_bitemporal_callbacks: false`' do
```
Comment (wata727) [inline spec/activerecord-bitemporal/callbacks_spec.rb:211]:
```suggestion
      context 'job title with option `ignore_bitemporal_callbacks: false`' do
```

File: spec/activerecord-bitemporal/callbacks_spec.rb [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+      before { employee.bitemporal_option_merge!(ignore_bitemporal_callbacks: false) }
+      it { expect { subject }.to change(employee, :log).from([]).to(%i[before_bitemporal_destroy after_bitemporal_destroy]) }
+    end
+
+    context 'with option `ignore_bitemporal_callbacks: true`' do
+      before { employee.bitemporal_option_merge!(ignore_bitemporal_callbacks: true) }
+      it { expect { subject }.not_to change(employee, :log) }
+    end
+
+    context 'has_one with `dependent: :destroy`' do
```
Comment (wata727) [inline spec/activerecord-bitemporal/callbacks_spec.rb:191]: Do we also need a test for nested attributes (`{ job_title_attributes: { _destroy: true } }`)?

File: lib/activerecord-bitemporal/callbacks.rb (hunk 5 above)
Comment (Dooor) [reply lib/activerecord-bitemporal/callbacks.rb]: I modified not to prepend Wrapper module.

File: spec/activerecord-bitemporal/callbacks_spec.rb (hunk 11 above)
Comment (Dooor) [reply spec/activerecord-bitemporal/callbacks_spec.rb]: I added nested attributes test case for bitemporal_destroy.

File: lib/activerecord-bitemporal.rb (hunk 4 above)
Comment (Dooor) [reply lib/activerecord-bitemporal.rb:127]: `Callbacks` module must be included by `ActiveModel` or `ActiveRecord` class because this module calls `define_callback` method when included.
`NoMethodError` will occur if `InstanceMthods` module include `Callbacks` module.

File: spec/activerecord-bitemporal/callbacks_spec.rb [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        context 'without option `ignore_bitemporal_callbacks`' do
+          it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_destroy after_bitemporal_destroy]) }
+        end
+
+        context 'employee with option `ignore_bitemporal_callbacks: true`' do
+          before { employee.bitemporal_option_merge!(ignore_bitemporal_callbacks: true) }
+          it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_destroy after_bitemporal_destroy]) }
+        end
+
+        context 'job_title with option `ignore_bitemporal_callbacks: false`' do
```
Comment (wata727) [inline spec/activerecord-bitemporal/callbacks_spec.rb:235]:
```suggestion
        context 'job_title with option `ignore_bitemporal_callbacks: true`' do
```

File: lib/activerecord-bitemporal/patches.rb (hunk 6 above)
Comment (Dooor) [reply lib/activerecord-bitemporal/patches.rb]: Behavior of whether or not a callbacks are executed depends on when to call `assign_nested_attributes`, so I thought it might cause bugs if `ignore_bitemporal_callbacks` option is copied to relation.

```rb
 # copied bitemporal_option to relation if call assign_attributes after set bitemporal_option
parent.bitemporal_option_merge!(ignore_bitemporal_callbacks: true)
parent.assign_attributes({ child_attributes: { ... } })
parent.save! # parent's and child's bitemporal_callbacks did not run

 # not copied bitemporal_option to relation if call assign_attributes before set bitemporal_option
parent.assign_attributes({ child_attributes: { ... } }) # copied bitemporal_option to relation
parent.bitemporal_option_merge!(ignore_bitemporal_callbacks: true)
parent.save! # child's bitemporal_callbacks are executed
```

But now I feel it would be more difficult that only `ignore_bitemporal_callbacks` option is treated specially. I removed this change.

File: lib/activerecord-bitemporal.rb (hunk 4 above)
Comment (wata727) [reply lib/activerecord-bitemporal.rb:127]: Fair enough. Nested `included` callbacks don't work...

File: lib/activerecord-bitemporal/patches.rb (hunk 6 above)
Comment (wata727) [reply lib/activerecord-bitemporal/patches.rb]: Yeah, you're right, this behavior is counter-intuitive and could be footgun. But this should be tackled as another issue.

**Review by wata727 — APPROVED** [review]
🥳

**Review by motsat — APPROVED** [review]
LGTM 👍
